keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -510,22 +516,36 @@ as the downstream connection within the dual concurrent `io.Copy` proxying gorou
 The upstream connection is the same SPDY connection to the container (through the
 Kubelet and CRI).
 
-### Pre-GA: Kubelet `StreamTranslatorProxy`
-
-The eventual plan is to incrementally transition all SPDY communication legs to WebSockets.
-After the WebSocket communication leg from `kubectl` to the API Server is proven
-to work, the next communication leg to transition is the one from the API Server to
-the Kubelet. Both the API Server and the Kubelet stream data messages using the
-`UpgradeAwareProxy`. Since the initial plan is to modify the `UpgradeAwareProxy`
-in the API Server to delegate to the `StreamTranslatorProxy`, it will be straightforward
-to transition this next communication leg by moving the integrated `StreamTranslatorProxy`
-from the API Server to the Kubelet.
-
-The final communication leg to transition from SPDY to WebSockets will be the one
-from Kubelet to the Container Runtimes. Since this communication happens within a
-node (using Unix domain sockets), this path is not as critical. But this effort
-will be more work, since it will require modifying not just Kubelet, but **all**
-Container Runtimes.
+### Proposal: Synthetic RBAC CREATE Authorization Check
+
+The transition to WebSockets requires changing the initial streaming upgrade request
+from a POST to a GET, as the WebSocket protocol specification
+([RFC 6455](https://www.rfc-editor.org/rfc/rfc6455#section-4.1)) mandates that the
+opening handshake must be an HTTP GET request. This has an unintended security
+consequence: RBAC policies that only grant the get verb, such as a typical read-only
+"viewer" role, now unexpectedly allow users to run kubectl exec, attach, and port-forward.
+To close this privilege escalation vector and restore the principle of least privilege,
+this proposal introduces a secondary, synthetic authorization check performed within the
+API Server. When a WebSocket upgrade request is received for the pods/exec, pods/attach,
+or pods/portforward subresources, the handler will perform an additional check to ensure
+the user also has the create verb permission for that specific subresource. This new
+authorization check will be controlled by a feature gate,
+`AuthorizePodWebsocketUpgradeCreatePermission`, which will be enabled by default to
```

Comment:
Will this feature gate be removed at some point in the future?. Or we'll keep it for a long time?.

Reply:
after a couple releases in enabled-by-default state, we'll transition it to locked on, then remove it ~3 releases after that